pkg/cloud/vsphere/services/loadbalancer/aws/loadbalancers.go

```diff
@@ -0,0 +1,315 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha2"
+	"sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha2/cloud"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services"
+)
+
+var (
+	sessionCache sync.Map
+)
+
+func sessionForRegion(region string) (*session.Session, error) {
+	s, ok := sessionCache.Load(region)
+	if ok {
+		return s.(*session.Session), nil
+	}
+
+	ns, err := session.NewSession(aws.NewConfig().WithRegion(region))
+	if err != nil {
+		return nil, err
+	}
+
+	sessionCache.Store(region, ns)
+	return ns, nil
+}
+
+// Service TODO
+type Service struct {
+	ELB elbv2iface.ELBV2API
+}
+
+// NewProvider TODO
+func NewProvider(config *cloud.LoadBalancerConfig) (services.LoadBalancerService, error) {
+	region := config.AwsProvider.Region
+	session, err := sessionForRegion(region)
+	if err != nil {
+		return nil, err
+	}
+	return &Service{
+		ELB: elbv2.New(session),
+	}, nil
+}
+
+func generateELBName(clusterName string) string {
+	return fmt.Sprintf("%s-endpoint", clusterName)
+}
+
+// Reconcile reconciles loadbalancing resources in aws
+func (svc *Service) Reconcile(vsphereCluster *infrav1.VSphereCluster, controlPlaneIPs []string) (infrav1.APIEndpoint, error) {
+	AwsProvider := vsphereCluster.Spec.LoadBalancerConfiguration.AwsProvider
+	targetGroupArn, err := svc.reconcileTargetGroup(vsphereCluster.Name, AwsProvider.VpcID, controlPlaneIPs)
+	if err != nil {
+		return infrav1.APIEndpoint{}, err
+	}
+
+	loadBalancerArn, loadBalancerDNS, err := svc.reconcileLoadBalancer(vsphereCluster.Name, AwsProvider.Subnets)
+	if err != nil {
+		return infrav1.APIEndpoint{}, err
+	}
+
+	listenerPort, err := svc.reconcileListeners(loadBalancerArn, targetGroupArn)
+	if err != nil {
+		return infrav1.APIEndpoint{}, err
+	}
+	apiEndpoint := infrav1.APIEndpoint{
+		Host: *loadBalancerDNS,
+		Port: int(*listenerPort),
+	}
+	return apiEndpoint, nil
+}
+
+func (svc *Service) reconcileTargetGroup(clusterName string, vpcID string, controlPlaneIPs []string) (*string, error) {
+	describeTargetGroupInput := &elbv2.DescribeTargetGroupsInput{}
+	describeTargetGroupInput.Names = append(describeTargetGroupInput.Names, aws.String(clusterName+"-controlPlane"))
+
+	describeTargetGroupOutput, err := svc.ELB.DescribeTargetGroups(describeTargetGroupInput)
+	desiredTargetGroupInput := &elbv2.CreateTargetGroupInput{
```

Comment:
Hi @yastij,

The order of the code here is confusing. It's necessary, but confusing. I think it makes sense in this case to group the declarations first. For example:

```golang
func (svc *Service) reconcileTargetGroup(clusterName string, vpcID string, controlPlaneIPs []string) (*string, error) {
	var (
		// targetGroupArn will store the ID of the discovered target group
		targetGroupArn *string

		// describeTargetGroupsInput is used to get a list of the target groups
		describeTargetGroupsInput = &elbv2.DescribeTargetGroupsInput{
			Names: []*string{aws.String(clusterName+"-controlPlane")},
		}

		// desiredTargetGroupInput describes the target group we want to find
		desiredTargetGroupInput = &elbv2.CreateTargetGroupInput{
			Name:                       aws.String(clusterName + "-controlPlane"),
			HealthCheckEnabled:         aws.Bool(true),
			HealthCheckIntervalSeconds: aws.Int64(int64(30 * time.Second.Seconds())),
			HealthCheckTimeoutSeconds:  aws.Int64(int64(10 * time.Second.Seconds())),
			HealthyThresholdCount:      aws.Int64(int64(3)),
			UnhealthyThresholdCount:    aws.Int64(int64(3)),
			Port:                       aws.Int64(int64(6443)),
			Protocol:                   aws.String(elbv2.ProtocolEnumTcp),
			TargetType:                 aws.String(string("ip")),
			VpcId:                      &vpcID,
		}
	)

	// Get a list of the target groups. If no target groups are discovered then
	// a new target group is created and its ARN is stored in targetGroupArn.
	describeTargetGroupsOutput, err := svc.ELB.DescribeTargetGroups(describeTargetGroupsInput)
	if err != nil {
		if IsNotFound(err) {
			createTargetGroupOutput, err := svc.ELB.CreateTargetGroup(desiredTargetGroupsInput)
			if err != nil {
				return nil, err
			}
			targetGroupArn = createTargetGroupOutput.TargetGroups[0].TargetGroupArn
		} else {
			return nil, err
		}
	}

	// If the target group already existed then make sure its ID is assigned to
	// targetGroupArn. If the target group didn't already exist, then its ID was
	// assigned to targetGroupArn when the target group was created above.
	if targetGroupArn == nil {
		targetGroupArn = describeTargetGroupsOutput.TargetGroups[0].TargetGroupArn
	}

	// Ensure the actual state of the target group matches the intended state of
	// target group. If the actual and intended states do not match then modify the
	// target group so its actual state matches the intended state.
	actualTargetGroupInput := convertTargetGroupOutputToInput(describeTargetGroupsOutput.TargetGroups[0])
	if !reflect.DeepEqual(actualTargetGroupInput, desiredTargetGroupInput) {
		modifyTargetGroup := &elbv2.ModifyTargetGroupInput{
			HealthCheckEnabled:         desiredTargetGroupInput.HealthCheckEnabled,
			HealthCheckIntervalSeconds: desiredTargetGroupInput.HealthCheckIntervalSeconds,
			HealthCheckTimeoutSeconds:  desiredTargetGroupInput.HealthCheckTimeoutSeconds,
			HealthyThresholdCount:      desiredTargetGroupInput.HealthyThresholdCount,
			UnhealthyThresholdCount:    desiredTargetGroupInput.UnhealthyThresholdCount,
		}
		if _, err := svc.ELB.ModifyTargetGroup(modifyTargetGroup); err != nil {
			return nil, err
		}
	}

	// If there is at least one control plane IP then register the IP addresses
	// to the target group.
	if len(controlPlaneIPs) > 0 {
		targetDescription := []*elbv2.TargetDescription{}
		for _, ipAddr := range controlPlaneIPs {
			targetDescription = append(targetDescription, &elbv2.TargetDescription{
				AvailabilityZone: aws.String(string("all")),
				Id:               aws.String(ipAddr),
				Port:             aws.Int64(int64(6443)),
			})
		}
		registerTargetInput := &elbv2.RegisterTargetsInput{
			TargetGroupArn: targetGroupArn,
			Targets:        targetDescription,
		}
		if _, err := svc.ELB.RegisterTargets(registerTargetInput); err != nil {
			return nil, err
		}
	}

	return targetGroupArn, nil
}
```